File: 백준2667 단지번호붙이기/복습/Main.java
import java.util.*;

public class Main {
    static int[][] map;
    static boolean[][] isVisited;
    static int N = 0; //정사각형 크기
    static int totalNumOfHouse = 0;
    static ArrayList<Integer> result = new ArrayList();
    private static int[] dx = {-1, 0, 1, 0};
    private static int[] dy = {0, -1, 0, 1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        map = new int[N + 2][N + 2];
        isVisited = new boolean[N + 2][N + 2];
        for (int i = 1; i <= N; i++) {
            String line = sc.next();
            for (int j = 1; j <= N; j++) {
                map[i][j] = line.charAt(j-1) - '0';
            }
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= N; j++) {
                if (!isVisited[i][j] && map[i][j] == 1)
                    bfs(i, j);
            }
        }
        System.out.println(totalNumOfHouse);
        Collections.sort(result);
        for (int i : result) {
            System.out.println(i);
        }
    }

    static void bfs(int x, int y) {
        int size = 1;
        Queue<Point> queue = new LinkedList<>();
        queue.offer(new Point(x, y));
        isVisited[x][y] = true;
        totalNumOfHouse++;
        while (!queue.isEmpty()) {
            Point point = queue.poll();
            for (int i = 0; i < 4; i++) {
                int x2 = point.x + dx[i];
                int y2 = point.y + dy[i];
                if (map[x2][y2] == 1 && !isVisited[x2][y2]) {
                    isVisited[x2][y2] = true;
                    size++;
                    queue.offer(new Point(x2, y2));
                }
            }
        }
        result.add(size);
    }

    static class Point {
        int x;
        int y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

}

File: 백준2667 단지번호붙이기/복습/Main.kt
import java.util.*

internal var map: Array<IntArray> = emptyArray()
internal var isVisited: Array<BooleanArray> = emptyArray()
internal var N = 0 //정사각형 크기
internal var totalNumOfHouse = 0
internal var result: ArrayList<Int> = ArrayList()
private val dx = intArrayOf(-1, 0, 1, 0)
private val dy = intArrayOf(0, -1, 0, 1)


fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    N = sc.nextInt()
    map = Array(N + 2) { IntArray(N + 2) }
    isVisited = Array(N + 2) { BooleanArray(N + 2) }
    for (i in 1..N) {
        val line = sc.next()
        for (j in 1..N) {
            map[i][j] = line[j - 1] - '0'
        }
    }

    for (i in 1..N) {
        for (j in 1..N) {
            if (!isVisited[i][j] && map[i][j] == 1)
                bfs(i, j)
        }
    }
    println(totalNumOfHouse)
    result.sort()
    for (i in result) {
        println(i)
    }
}

internal fun bfs(x: Int, y: Int) {
    var size = 1
    val queue = LinkedList<Point>()
    queue.offer(Point(x, y))
    isVisited[x][y] = true
    totalNumOfHouse++
    while (!queue.isEmpty()) {
        val point = queue.poll()
        for (i in 0..3) {
            val x2 = point.x + dx[i]
            val y2 = point.y + dy[i]
            if (map[x2][y2] == 1 && !isVisited[x2][y2]) {
                isVisited[x2][y2] = true
                size++
                queue.offer(Point(x2, y2))
            }
        }
    }
    result.add(size)
}

data class Point(var x: Int, var y: Int)

